metaflow/plugins/kfp/kfp.py

```diff
@@ -1,316 +1,395 @@
+            "--task-id %s" % task_id,
+            # Since retries are handled by KFP Argo, we can rely on
+            # {{retries}} as the job counter.
+            # '--retry-count {{retries}}',  # TODO: test verify, should it be -1?
+            "--max-user-code-retries %d" % user_code_retries,
+            "--input-paths %s" % paths,
+        ]
+
+        if any(self.graph[n].type == "foreach" for n in node.in_funcs):
+            # We set the `METAFLOW_SPLIT_INDEX` through JSONPath-foo
+            # to pass the state from the parent DynamoDb state for for-each.
+            step.append("--split-index $METAFLOW_SPLIT_INDEX")  # TODO: get from KFP param
+        if self.namespace:
+            step.append("--namespace %s" % self.namespace)
+        cmds.append(" ".join(entrypoint + top_level + step))
+        return " && ".join(cmds)
+
+    def create_kfp_pipeline_from_flow_graph(self):
+        import kfp
+        from kfp import dsl
+
+        base_image = "ssreejith3/mf_on_kfp:python-curl-git"  # TODO: environment variable
+        step_to_command_template_map = self.create_command_templates_from_graph()
+
+        # Container op that corresponds to a step defined in the Metaflow flowgraph.
+        step_op = kfp.components.func_to_container_op(step_op_func, base_image=base_image)
+
+        @dsl.pipeline(
+            name="MF on KFP Pipeline",
+            description="Pipeline defining KFP equivalent of the Metaflow Flow."
+            "with branch and join nodes",
+        )
+        def kfp_pipeline_from_flow():
+            kfp_run_id = "kfp-" + dsl.RUN_ID_PLACEHOLDER
+
+            def build_kfp_dag(start: DAGNode):
+                visited = {}
+
+                def dfs(node: DAGNode, context, index=None):
+                    visited[node.name] = step_op(
+                        step_to_command_template_map[node.name], kfp_run_id, context, index=index
+                    ).set_display_name(node.name)
+
+                    if node.type == "foreach":
+                        with kfp.dsl.ParallelFor(
+                            visited[node.name].outputs["split_indexes"]
+                        ) as index:
+                            for step in node.out_funcs:
+                                dfs(
+                                    self.graph[step],
+                                    visited[node.name].outputs["task_out_dict"],
+                                    index,
+                                )
+                        # TODO: the join
+                        # visited[node.name].outputs["task_out_dict"]
+                    else:
+                        for step in node.out_funcs:
+                            if step not in visited:
+                                dfs(self.graph[step], visited[node.name].outputs["task_out_dict"])
+
+                            visited[step].after(visited[node.name])
+
+                dfs(start, {})
+
+            build_kfp_dag(self.graph["start"])
```

Comment:
Do you need the `build_kfp_dag` inner function, doesn't seem to be doing much at the moment at least, can we remove it to simplify this method?